// adonis-react/lib/src/React.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareReact = void 0;
const react_1 = __importDefault(require("react"));
const server_1 = require("react-dom/server");
const adonisContext_1 = require("./adonisContext");
function prepareReact(ctx, app, Route) {
    const instance = new ReactInstance();
    instance.share({
        app,
        ctx,
        request: ctx.request,
        params: ctx.params,
        route: ctx.route,
        makeUrl(...args) {
            return Route.makeUrl(...args);
        },
        makeSignedUrl: (...args) => {
            return Route.makeSignedUrl(...args);
        },
    });
    return instance;
}
exports.prepareReact = prepareReact;
class ReactInstance {
    constructor() {
        this.context = {};
    }
    share(obj) {
        Object.assign(this.context, obj);
    }
    render(Component, props) {
        const html = server_1.renderToStaticMarkup(
        // @ts-expect-error Context is filled outside of this class
        react_1.default.createElement(adonisContext_1.AdonisContextProvider, { value: this.context },
            react_1.default.createElement(Component, Object.assign({}, props))));
        return `<!DOCTYPE html>\n${html}`;
    }
}
